static-analysis/teammates-macker.xml

```diff
@@ -1,45 +1,340 @@
+                    </allow>
+                    <allow>
+                        <to>
+                            <!-- allow dependency to inner class -->
+                            <include class="${testCasesPackage}.**$**" />
+                            <!-- allow dependency to base class -->
+                            <include class="${testCasesPackage}.**Base**" />
+                        </to>
+                    </allow>
+                    <allow>
+                        <!-- allow dependency from Feedback*QuestionUiTest to FeedbackQuestionUiTest -->
+                        <from class="${testCasesPackage}.browsertests.Feedback*QuestionUiTest" />
+                        <to class="${testCasesPackage}.browsertests.FeedbackQuestionUiTest" />
+                    </allow>
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="TestDriverComponent: Only UI tests can access page object classes">
+            <access-rule>
+                <deny>
+                    <from class="teammates.test.**" />
+                    <to class="teammates.test.pageobjects.**" />
+                    <allow>
+                        <from>
+                            <include class="teammates.test.pageobjects.**" />
+                            <include class="teammates.test.cases.browsertests.**" />
+                        </from>
+                    </allow>
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="TestDriverComponent: Test driver can only be accessed from test package">
+            <access-rule>
+                <deny>
+                    <from class="teammates.**" />
+                    <to class="teammates.test.driver.**" />
+                    <allow>
+                        <from>
+                            <include class="teammates.test.**" />
+                            <include class="teammates.client.**" />
+                        </from>
+                    </allow>
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="TestDriverComponent: Only certain test cases can test storage:entity">
+            <access-rule>
+                <deny>
+                    <from class="teammates.test.**" />
+                    <to class="teammates.storage.entity.**" />
+                    <allow>
+                        <from class="teammates.test.cases.datatransfer.**" />
+                    </allow>
+                    
+                    <!-- Violations that need fixing -->
+                    <allow>
+                        <from class="teammates.test.cases.logic.StudentsLogicTest" />
+                        <to class="teammates.storage.entity.CourseStudent" />
+                    </allow>
+                    <allow>
+                        <from class="teammates.test.cases.storage.FeedbackResponseCommentsDbTest" />
```

Comment:
Actually this might be allowed, because the Storage layer can actually access entities. You can just allow `teammates.test.cases.storage.**` for now; we'll review this later if needed.